### PR TITLE
Added integration test and test for weather model

### PR DESCRIPTION
**Pull request by RN35**
## JIRA: 
Link : [LAB-24](https://oregonstate-innovationlab.atlassian.net/browse/LAB-24) # this is just a dummy link
## Features:
- Added a narrow integration test for `WeatherService`
- Added unit test for `Weather` model

## What is not handled:
N/A

## Assumptions:
- The client assumes that a mocker server is running with following specifications:
Endpoint: http://localhost:3000/data/2.5/weather
Response:
```
{"coord":{"lon":-123.262,"lat":44.5646},"weather":[{"id":801,"main":"Clouds","description":"few clouds","icon":"02n"}],"base":"stations","main":{"temp":58,"feels_like":49.98,"temp_min":48.09,"temp_max":55.45,"pressure":1023,"humidity":78},"visibility":10000,"wind":{"speed":0,"deg":0},"clouds":{"all":20},"dt":1641958461,"sys":{"type":2,"id":2040223,"country":"US","sunrise":1641916079,"sunset":1641948820},"timezone":-28800,"id":5720727,"name":"Corvallis","cod":200}
```

## Note for reviewer:
- Manipulate the json response from the mock server to see the tests failing

### CODE REVIEW

File: MyLibrary/Tests/MyLibraryTests/MyLibraryTests.swift [hunk 1]
```diff
@@ -68,5 +69,38 @@ final class MyLibraryTests: XCTestCase {
...
+        let decoder = JSONDecoder()
+        let data = Data("{\"main\": {\"temp\":58}}".utf8)
+        do {
+            let container = try decoder.decode(Weather.self, from: data)
+            XCTAssertNotNil(container)
+            XCTAssert(container.main.temp == 58.0)
+        } catch {
+            XCTFail("Failed to decode temp from json")
+        }
+    }
```
Comment (wbraynen) [inline MyLibrary/Tests/MyLibraryTests/MyLibraryTests.swift]: You don't need a try-catch block.  You could instead just have the test function throw:
```
func testWeatherModelForValidTemperature() throws {`
```
and the test will automatically fail if it throws.

Also, although not everyone is a fan, I would encourage you to stick with "Given / When / Then" (see the test above for an example).

File: MyLibrary/Tests/MyLibraryTests/MyLibraryTests.swift [hunk 2]
```diff
@@ -68,5 +69,38 @@ final class MyLibraryTests: XCTestCase {
...
+        do {
+            let _ = try decoder.decode(Weather.self, from: data)
+            XCTFail("Should not have been possible to decode temp for invalid input")
+        } catch {
+            XCTAssert(true == true)
+        }
+    }
+    
+    // Integration Tests
+    func testWeatherServiceReturnsCorrectTemperature() async {
```
Comment (wbraynen) [inline MyLibrary/Tests/MyLibraryTests/MyLibraryTests.swift]: If you put the integration tests in the same test suite (and here they are part of the same test file to boot), then you won't be able to launch integration tests up on the CI without explicitly disabling this test in the scheme.

So, instead, I would recommend creating a separate test suite.  You can create a second unit-test suite and rename it to "Integration Tests".  I would then see how to trigger this or that test suite from the command line.

File: MyLibrary/Tests/MyLibraryTests/MyLibraryTests.swift [hunk 3]
```diff
@@ -68,5 +69,38 @@ final class MyLibraryTests: XCTestCase {
...
+        let weatherServiceObj = WeatherServiceImpl()
```
Comment (wbraynen) [inline MyLibrary/Tests/MyLibraryTests/MyLibraryTests.swift]: A nitpick, but I wouldn't call objects "objects".  We know they are objects.  I would just call it a `weatherService` instead of a `weatherServiceObj`.

File: MyLibrary/Sources/MyLibrary/WeatherService.swift [hunk 4]
```diff
@@ -4,15 +4,21 @@ public protocol WeatherService {
...
 
-class WeatherServiceImpl: WeatherService {
-    let url = "https://api.openweathermap.org/data/2.5/weather?q=corvallis&units=imperial&appid=<INSERT YOUR API KEY HERE>"
+enum ServiceDomain : String {
+  case actual = "https://api.openweathermap.org"
+  case mock = "http://localhost:3000"
+}
 
+class WeatherServiceImpl: WeatherService {
+    let url = "\(ServiceDomain.mock.rawValue)/data/2.5/weather?q=corvallis&units=imperial&appid=INSERT_YOUR_API_KEY_HERE"
```
Comment (wbraynen) [inline MyLibrary/Sources/MyLibrary/WeatherService.swift]: While you are at it, why not have a constant called `apiKey` that you can string-interpolate here?  You are already using _string interpolation_ to insert the base url.

File: MyLibrary/Tests/MyLibraryTests/MyLibraryTests.swift [hunk 5]
```diff
@@ -68,5 +69,38 @@ final class MyLibraryTests: XCTestCase {
...
+        do{
+            let temp = try await weatherServiceObj.getTemperature()
+            XCTAssert(temp is Int)
+        } catch {
+            XCTFail("Failed to hit weather service")
+        }
+    }
```
Comment (wbraynen) [inline MyLibrary/Tests/MyLibraryTests/MyLibraryTests.swift]: I would add some tests for the model too.  That way, if the model breaks (e.g. someone puts a PR that breaks it), you'll know right away.

File: MyLibrary/Sources/MyLibrary/WeatherService.swift [hunk 6]
```diff
@@ -24,7 +30,7 @@ class WeatherServiceImpl: WeatherService {
     }
 }
 
-private struct Weather: Decodable {
+struct Weather: Decodable {
```
Comment (wbraynen) [inline MyLibrary/Sources/MyLibrary/WeatherService.swift:35]: Why this change if you are not testing the `Weather` model?

File: MyLibrary/Tests/MyLibraryTests/MyLibraryTests.swift (hunk 1 above)
Comment (RN35) [reply MyLibrary/Tests/MyLibraryTests/MyLibraryTests.swift]: Got it! Addressed.

File: MyLibrary/Tests/MyLibraryTests/MyLibraryTests.swift (hunk 2 above)
Comment (RN35) [reply MyLibrary/Tests/MyLibraryTests/MyLibraryTests.swift]: Understood. Organized into separate folders. One way to run only specific tests using the `swift` command would be `swift test --filter "WeatherServiceIntegrationTests"`. 
image

File: MyLibrary/Tests/MyLibraryTests/MyLibraryTests.swift (hunk 3 above)
Comment (RN35) [reply MyLibrary/Tests/MyLibraryTests/MyLibraryTests.swift]: Addressed.

File: MyLibrary/Sources/MyLibrary/WeatherService.swift (hunk 4 above)
Comment (RN35) [reply MyLibrary/Sources/MyLibrary/WeatherService.swift]: Addressed.

File: MyLibrary/Tests/MyLibraryTests/MyLibraryTests.swift (hunk 5 above)
Comment (RN35) [reply MyLibrary/Tests/MyLibraryTests/MyLibraryTests.swift]: Test already present 🙂 ref: https://github.com/RN35/cs561-swift/pull/1/files#diff-00e50ae2d22c75d777b90ff4db48abf4aa63e59b5846d12588ffc59c63109894R26

File: MyLibrary/Sources/MyLibrary/WeatherService.swift (hunk 6 above)
Comment (RN35) [reply MyLibrary/Sources/MyLibrary/WeatherService.swift:35]: Unit test for `Weather` model present 🙂
